test/jdk/java/util/zip/ZipCoding.java

```diff
@@ -23,111 +23,184 @@
+                // ZipOutputStream sets the 'Language encoding flag' when writing using UTF-8
+                // UTF-8 should be used for decoding, regardless of the opening charset
 
+                // UTF-8 with Japanese characters, opened with MS932
+                Arguments.of("utf-8", "MS932",
+                        "\u4e00\u4e01",
+                        "\uff67\uff68\uff69\uff6a\uff6b\uff6c"),
+
+                // UTF-8 with characters in latin1 range, opened with iso-8859-1
+                Arguments.of("utf-8", "iso-8859-1",
+                        "\u00e4\u00fc",
+                        "German Umlaut \u00fc in comment"),
+                // UTF-8 with surrogate pairs, opened with MS932
+                Arguments.of("utf-8", "MS932",
+                        "Surrogate\ud801\udc01",
+                        "Surrogates \ud800\udc00 in comment")
+        );
     }
 
-    static void testZipInputStream(InputStream is, Charset cs,
-                                   String name, String comment, byte[] bb)
-        throws Exception
-    {
-        try (ZipInputStream zis = new ZipInputStream(is, cs)) {
+    /**
+     * Verify that ZipInputStream decodes entry names and comments
+     * using the charset provided to its constructor, or that it decodes
+     * using UTF-8 when the 'Language encoding flag' is set
+     *
+     * @param writeCharset the charset to use for ZipOutputStream when producing the ZIP
+     * @param readCharset the charset to use when opening the ZipInputStream
+     * @param name the entry name
+     * @param comment the entry comment (not read by ZipInputStream)
+     *
+     * @throws IOException if an unexpected IOException occurs
+     */
+    @ParameterizedTest
+    @MethodSource("charsetsAndNames")
+    public void testZipInputStream(String writeCharset,
+                                   String readCharset,
+                                   String name,
+                                   String comment) throws IOException {
+
+        byte[] zip = createZIP(writeCharset, name, comment);
+
+        try (InputStream in = new ByteArrayInputStream(zip);
+             ZipInputStream zis = new ZipInputStream(in, Charset.forName(readCharset))) {
             ZipEntry e = zis.getNextEntry();
-            if (e == null || ! name.equals(e.getName()))
-                throw new RuntimeException("ZipIS name doesn't match!");
-            byte[] bBuf = new byte[bb.length << 1];
-            int n = zis.read(bBuf, 0, bBuf.length);
-            if (n != bb.length ||
-                !Arrays.equals(bb, Arrays.copyOf(bBuf, n))) {
-                throw new RuntimeException("ZipIS content doesn't match!");
-            }
+            assertNotNull(e);
+            assertEquals(name, e.getName());
+            assertNull(e.getComment()); // No comment in the LOC header
+            assertArrayEquals(ENTRY_DATA, zis.readAllBytes(), "ZipIS content doesn't match!");
```

Comment:
Minor Nit, but perhaps changes the "ZipIS content..." message to make it clearer as it is not what is clear what ZipIS is...

Reply:
These messages were mostly carried over. I updated this one to "ZIP entry data does not match". Also updated a few other assertion messages.